libvirt_volume: Error while uploading source os-image.qcow2: unexpected EOF

#  Version Reports:

### Distro  version of host:
Ubuntu
``` openSUSE 42.2/ Centos7/ Ubuntu.. ```
Ubuntu 18.04
### Terraform Version Report
(Provided by running `terraform -v`.)
0.11.7
### Libvirt version
libvirtd (libvirt) 4.0.0
```virsh --version```

I hit a smilar error just now when using a single `bridge = "..."` statement inside a `network_interface { }`.

```hcl
resource "libvirt_domain" "coreos" {
  # basics
  name   = "coreos-${count.index}"
  memory = "1024"
  vcpu   = 2

  disk {
    volume_id = "${libvirt_volume.coreos.*.id[count.index]}"
  }

  network_interface {
    bridge = "kvm-bridge"
  }

  boot_device {
    dev = ["hd", "network"]
  }

  # needs qemu >= 2.6
  coreos_ignition = "${libvirt_ignition.ignition.id}"
  count           = "${var.cluster_size}"
}
```

It crashed Terraform entirely:
```
panic: runtime error: invalid memory address or nil pointer dereference
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xc518b7]
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: goroutine 75 [running]:
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/libvirt.getDomainInterfacesFromNetworks(0x0, 0x0, 0xc420311048, 0x6, 0xc420311078, 0x3, 0xc420311080, 0xc4203110a8, 0x8, 0xc420318570, ...)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/terraform-provider-libvirt-0.4.1/_build/src/github.com/dmacvicar/terraform-provider-libvirt/libvirt/domain.go:209 +0x157
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/libvirt.domainGetIfacesInfo(0x7fa6080014f0, 0x0, 0x0, 0xc420311048, 0x6, 0xc420311078, 0x3, 0xc420311080, 0xc4203110a8, 0x8, ...)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/terraform-provider-libvirt-0.4.1/_build/src/github.com/dmacvicar/terraform-provider-libvirt/libvirt/domain.go:167 +0x196
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/libvirt.resourceLibvirtDomainRead(0xc420162e00, 0xe2d380, 0xc42031f220, 0x0, 0x0)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/terraform-provider-libvirt-0.4.1/_build/src/github.com/dmacvicar/terraform-provider-libvirt/libvirt/resource_libvirt_domain.go:760 +0x1b63
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/libvirt.resourceLibvirtDomainCreate(0xc420162e00, 0xe2d380, 0xc42031f220, 0x0, 0x0)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/terraform-provider-libvirt-0.4.1/_build/src/github.com/dmacvicar/terraform-provider-libvirt/libvirt/resource_libvirt_domain.go:503 +0x11d7
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc42009e540, 0xc420308640,0xc420313a40, 0xe2d380, 0xc42031f220, 0x1, 0xc42031a4e0, 0xc420462c00)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/terraform-provider-libvirt-0.4.1/_build/src/github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:227 +0x364
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc42009e850, 0xc4203085f0,0xc420308640, 0xc420313a40, 0x7fa62bbdd6c8, 0x0, 0x18)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/terraform-provider-libvirt-0.4.1/_build/src/github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:283 +0xa4
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc4204189e0, 0xc4203139e0, 0xc4204a7970, 0x0, 0x0)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /home/abuild/rpmbuild/BUILD/terraform-provider-libvirt-0.4.1/_build/src/github.com/dmacvicar/terraform-provider-libvirt/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:527 +0x57
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt: reflect.Value.call(0xc4200613e0, 0xc420426370, 0x13, 0x1018fc8, 0x4, 0xc420346f20, 0x3, 0x3, 0x0, 0x0, ...)
2018-08-02T16:25:55.360+0200 [DEBUG] plugin.terraform-provider-libvirt:         /usr/lib/golang/src/reflect/value.go:434 +0x905
2018-08-02T16:25:55.361+0200 [DEBUG] plugin.terraform-provider-libvirt: reflect.Value.Call(0xc4200613e0, 0xc420426370, 0x13, 0xc420367f20, 0x3, 0x3, 0x0, 0x0, 0x0)
2018-08-02T16:25:55.361+0200 [DEBUG] plugin.terraform-provider-libvirt:         /usr/lib/golang/src/reflect/value.go:302 +0xa4
2018-08-02T16:25:55.361+0200 [DEBUG] plugin.terraform-provider-libvirt: net/rpc.(*service).call(0xc420499a00, 0xc420084140, 0xc4204221d8, 0xc420316080, 0xc420112d00, 0xe319c0, 0xc4203139e0, 0x16, 0xe31a00, 0xc4204a7970, ...)
2018-08-02T16:25:55.361+0200 [DEBUG] plugin.terraform-provider-libvirt:         /usr/lib/golang/src/net/rpc/server.go:381 +0x142
2018-08-02T16:25:55.361+0200 [DEBUG] plugin.terraform-provider-libvirt: created by net/rpc.(*Server).ServeCodec
2018-08-02T16:25:55.361+0200 [DEBUG] plugin.terraform-provider-libvirt:         /usr/lib/golang/src/net/rpc/server.go:475 +0x36b
2018/08/02 16:25:55 [TRACE] root: eval: *terraform.EvalWriteState
2018/08/02 16:25:55 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2018/08/02 16:25:55 [TRACE] root: eval: *terraform.EvalIf
2018/08/02 16:25:55 [TRACE] root: eval: *terraform.EvalWriteState
2018/08/02 16:25:55 [TRACE] root: eval: *terraform.EvalWriteDiff
2018/08/02 16:25:55 [TRACE] root: eval: *terraform.EvalApplyPost
2018-08-02T16:25:55.365+0200 [DEBUG] plugin: plugin process exited: path=/home/ansemjo/.terraform.d/plugins/terraform-provider-libvirt
2018/08/02 16:25:55 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* libvirt_domain.coreos: unexpected EOF
2018/08/02 16:25:55 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* libvirt_domain.coreos: unexpected EOF
2018/08/02 16:25:55 [TRACE] [walkApply] Exiting eval tree: libvirt_domain.coreos
2018/08/02 16:25:55 [TRACE] dag/walk: upstream errored, not walking "provider.libvirt (close)"
2018/08/02 16:25:55 [TRACE] dag/walk: upstream errored, not walking "meta.count-boundary (count boundary fixup)"
2018/08/02 16:25:55 [TRACE] dag/walk: upstream errored, not walking "root"
2018/08/02 16:25:55 [TRACE] Preserving existing state lineage "202f52d1-e03a-651d-3c61-039f4658aef7"
2018/08/02 16:25:55 [DEBUG] plugin: waiting for all plugin processes to complete...
2018-08-02T16:25:55.372+0200 [WARN ] plugin: error closing client during Kill: err="connection is shut down"
2018-08-02T16:25:55.374+0200 [DEBUG] plugin: plugin process exited: path=/home/ansemjo/terraform-project/.terraform/plugins/linux_amd64/terraform-provider-ignition_v1.0.1_x4
```

Can't say for sure which version of the provider I was using though:

```
$ terraform -v
Terraform v0.11.7
+ provider.ignition v1.0.1
+ provider.libvirt (unversioned)
```

Updating the provider by compiling it from source seems to have fixed the problem. No more crashes so far ...

```
$ git -C ~/go/src/github.com/dmacvicar/terraform-provider-libvirt/ rev-parse HEAD
84823f8f34e184e31899b61710843a2b9af215c6
```

reproducing with latest provider
```
* libvirt_volume.centos7_volume: Error while uploading source http://schnell.nue.suse.com/sumaform-images/centos7.qcow2: unexpected EOF
2018/08/20 20:08:36 [ERROR] root.base: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* libvirt_volume.centos7_volume: Error while uploading source http://schnell.nue.suse.com/sumaform-images/centos7.qcow2: unexpected EOF
```

--- a/libvirt/utils_volume.go
+++ b/libvirt/utils_volume.go
@@ -160,7 +160,12 @@ func isQCOW2Header(buf []byte) (bool, error) {
 
 func (i *httpImage) Import(copier func(io.Reader) error, vol libvirtxml.StorageVolume) error {
 	client := &http.Client{}
-	req, _ := http.NewRequest("GET", i.url.String(), nil)
+	req, err := http.NewRequest("GET", i.url.String(), nil)
+
+	if err != nil {
+		log.Printf("[DEBUG:] Error creating new request for source url %s: %s", i.url.String(), err)
+		return fmt.Errorf("Error while downloading %s: %s", i.url.String(), err)
+	}
 
 	if vol.Target.Timestamps != nil && vol.Target.Timestamps.Mtime != "" {
 		req.Header.Set("If-Modified-Since", timeFromEpoch(vol.Target.Timestamps.Mtime).UTC().Format(http.TimeFormat))
@@ -175,7 +180,7 @@ func (i *httpImage) Import(copier func(io.Reader) error, vol libvirtxml.StorageV
 	if response.StatusCode == http.StatusNotModified {
 		return nil
 	}
-
+	log.Printf("[DEBUG]: url resp status code %s\n", response.Status)
 	return copier(response.Body)
 }
 
@@ -212,13 +217,21 @@ func newCopier(virConn *libvirt.Connect, volume *libvirt.StorageVol, size uint64
 			stream.Free()
 		}()
 
-		volume.Upload(stream, 0, size, 0)
+		if err := volume.Upload(stream, 0, size, 0); err != nil {
+			stream.Abort()
+			return fmt.Errorf("Error while uploading volume %s", err)
+		}
 
 		sio := NewStreamIO(*stream)
 
 		bytesCopied, err = io.Copy(sio, src)
+		// if we get unexpected EOF this mean that connection was closed suddently from server side
+		// the problem is not on the plugin but on server hosting currupted images
+		if err == io.ErrUnexpectedEOF {
+			return fmt.Errorf("Error: transfer was unexpectedly closed from the server while downloading. Please try again later or check the server hosting sources")
+		}
 		if err != nil {
-			return err
+			return fmt.Errorf("Error while copying source to volume %s", err)
 		}
 		log.Printf("%d bytes uploaded\n", bytesCopied)
 		return nil

